src/shogun/classifier/Perceptron.cpp

```diff
@@ -77,42 +52,30 @@ bool CPerceptron::train_machine(CFeatures* data)
 		w.set_const(1.0 / num_feat);
 		bias=0;
 	}
-	else
-	{
-		w = get_w();
-	}
-	//loop till we either get everything classified right or reach max_iter
-	while (!converged && iter < max_iter)
+}
+
+void CPerceptron::iteration(CDotFeatures* f, CLabels* l)
+{
+	m_complete = true;
+	SGVector<float64_t> w = get_w();
+	SGVector<float64_t> output(f->get_num_vectors());
+	auto iter_train_labels = binary_labels(l)->get_int_labels().begin();
+	auto iter_output = output.begin();
+	for (const auto& v : DotIterator(f))
```

Comment:
Aren't we doing a sort of batch learning here? This ``iteration`` method will run over all the features so, if we premature stop ``continue_train`` and then we call it again, the `iteration` method will just run over all the features `f` from the start. This would work if we provided to ``iteration`` a set of new features/labels each time ``continue_train`` is called. @karlnapf @shubham808

Reply:
for Perceptron this should be ok as we are touching *all* data in *every* iteration. So there is no problem with continued algorithms starting to touch the data from beginning again